---
layout: default
title: About
permalink: /about/
---

# About me

## EDUCATION
- PhD Candidate in Bioinformatics and Integrative Genomics, Harvard University, Division of Medical Science (June 2013 - Present)
- Bachelors of Science in Biomedical Engineering and Applied Math, Johns Hopkins University, Whiting School of Engineering (Aug. 2009 - May 2013)
- High School Diploma, Magnet Program, Montgomery Blair High School (Aug. 2006 - May 2009)



## WORK AND RESEARCH EXPERIENCE
- Graduate Student, Mentored by Peter Kharchenko, Department of Biomedical Informatics, Harvard Medical School (July 2013 - Present)
- Rotation Student, Mentored by Nir Hacohen and Catherine Wu, Broad Institute (Sep. 2013 - Nov. 2013)
- Summer Research Intern, Mentored by Shamil Sunyaev, Bioinformatics and Integrative Genomics Program, Harvard-MIT HST and i2b2 (June 2012 - Aug. 2012)
- Research Intern, Mentored by Rachel Karchin, Institute of Computational Medicine, Johns Hopkins University (Aug. 2009 - May 2013)
- Summer Research Intern, Mentored by Paul Meltzer and Liang Cao, Center for Cancer Research, National Cancer Institute, National Institutes of Health (June 2008 - Oct. 2008)

## TEACHING EXPERIENCE
- Lead Instructor, Intro to Computer Programming through Web Development with Social Entrepreneurship (Sept. 2016 - Present)
- Teaching Assistant, Summer Institute in Biomedical Informatics DBMI/HST (June 2016 - Aug. 2016)
- Lead Instructor, Intro to Computer Programming for Future Scientists and Engineers The Innovation Institute (Sept. 2015 - June 2016)
- Teaching Fellow, Single Cell Workshop 2015 Harvard Stem Cell Institute (Nov. 2015)
- Teaching Fellow, Single Cell Workshop 2014 Harvard Stem Cell Institute (Oct. 2014)
- Teaching Assistant, Introduction to Optimization 550.361 Johns Hopkins University (Aug. 2012 - Dec. 2012)

## SELECT INVITED TALKS AND PRESENTATIONS
- **Fan J.** Applying single cell transcriptomics - unraveling the complexity of the developing human
brain. Festival of Genomics, Single Cell Genomics, 2016 (Talk)
- **Fan J.** Do art like a science. Do science like an art. TEDxJHU, Instructions Not Included, 2016
(Talk) 
- **Fan J**, Wang L, Brooks AN, Wan Y, Neuberg D, Rassenti K, Ghia E, Kipps T, Brown JR, Li S,
Livak KJ, Meyerson MM, Kharchenko PV, Wu CJ Comprehensive Bulk and Single Cell
Transcriptomic Characterization of SF3B1 Mutation Reveals its Pleiotropic Eects in Chronic
Lymphocytic Leukemia. American Society of Hematology, 57th Annual Meeting and Exhibition,
2015 (Poster)
- **Fan J**, Salathia N, Liu R, Kaeser G, Yung Y, Herman J, Kaper F, Fan JB, Zhang K, Chun J, and
Kharchenko PV. PAGODA - Pathway and gene set overdispersion analysis characterizes single cell
transcriptional heterogeneity. Cold Spring Harbor. Single Cell Analysis Meeting, 2015 (Talk)
- **Fan J**, Lee HO, Lee S, Lee A, Park WY, Park PJ, Kharchenko PV. Hierarchical Bayesian
Approach for CNV Detection from Single Cell RNA-seq Data. National Institutes of Health, Single
Cell Investigators Meeting, 2015 (Poster)
- **Fan J**, Salathia N, Liu R, Yung Y, Fan JB, Chun J, Zhang K, Kharchenko PV. Single Cell
Dierential Expression Identies Neural Progenitor Subpopulations in the Developing Mouse Brain.
National Institutes of Health, Single Cell Investigators Meeting, 2014 (Poster)
- **Fan J**, Karchin R. Computational Assessment of the Utility of Limiting Orthologous Sequence
Depth in Mutation Impact Prediction Performance. International Congress of Human
Genetics/American Society of Human Genetics Conference, 2011 (Poster)

## SELECT PUBLICATIONS ([ORCID](orcid.org/0000-0002-0212-5451))
- **Fan J**, Salathia N, Liu R, Kaeser G, Yung Y, Herman J, Kaper F, Fan JB, Zhang K, Chun J, and
Kharchenko PV. Characterizing transcriptional heterogeneity through pathway and gene set
overdispersion analysis. Nature Methods 2016 Mar;13(3):241-4. [doi:10.1038/nmeth.3734](http://www.nature.com/nmeth/journal/v13/n3/abs/nmeth.3734.html)
- Zhang X, Chen MH, Wu X, Kodani A, **Fan J**, Doan R, Ozawa M, Ma J, Yoshida N, Reiter JF, et al. Cell-Type-Specific Alternative Splicing Governs Cell Fate in the Developing Cerebral Cortex. Cell. 2016;166 (5) :1147-1162.e15. [doi:10.1016/j.cell.2016.07.025](http://www.sciencedirect.com/science/article/pii/S0092867416309321)
- Burger JA*, Landau DA*, Taylor-Weiner A*, Bozic I*, Zhang H*, Sarosiek K, Wang L, Stewart C,
**Fan J**, Hoellenriegel J, Sivina M, Dubuc AM, Fraser C, Han Y, Li S, Livak KJ, Zou L, Wan Y,
Konoplev S, Sougnez C, Brown JR, Abruzzo LV, Carter SL, Keating MJ, Davids MS, Wierda WG,
Cibulskis K, Zenz T, Werner L, Dal Cin P, Kharchencko P, Neuberg D, Kantarjian H, Lander E,
Gabriel S, O'Brien S, Letai A, Weitz DA, Nowak MA, Getz G, Wu CJ. Clonal evolution in patients
with chronic lymphocytic leukaemia developing resistance to BTK inhibition. Nature
Communications 2016, May 20. [doi:10.1038/ncomms11589](http://www.nature.com/articles/ncomms11589)
- Landau DA, Clement K, Ziller MJ, Boyle P, **Fan J**, Gu H, Stevenson K, Sougnez C, Wang L, Li S, Kotliar D, Zhang W, Ghandi M, Garraway L, Fernandes SM, Livak KJ, Gabriel S, Gnirke A, Lander ES, Brown JR, Neuberg D, Kharchenko PV, Hacohen N, Getz G, Meissner A and Wu CJ. Locally disordered methylation forms the basis of intratumor methylome variation in chronic lymphocytic leukemia. Cancer Cell 2014, Dec 8; 26(6):813-25. [doi:10.1016/j.ccell.2014.10.012](http://www.sciencedirect.com/science/article/pii/S1535610814004164)

## FELLOWSHIPS
- Ruth L. Kirschstein Predoctoral Individual National Research Service Award F31 Grant recipient,
CA206236-01 (2016)
- National Science Foundation Graduate Research Fellowship Program awardee, DGE144152 (2013)
- Department of Defense American Society for Engineering Education National Defense Science and
Engineering Graduate Fellowship awardee (2013)

## SELECT AWARDS AND HONORS
- ASH Abstract Achievement Award Recipient (2015)
- Provost's Undergraduate Research Award Winner (2012)
- Intel Science Talent Search Semi-Finalist (2009)
- Siemens Competition in Math, Science and Technology Semi-Finalist (2008)
